src/builder/QuarkBuilder.sol

```diff
@@ -1092,7 +1213,35 @@ contract QuarkBuilder {
             balanceOnChain += counterpartBalance;
         }
 
-        return balanceOnChain < amountNeededOnDstChain;
+        return balanceOnChain;
+    }
+
+    function getAmountNeededOnChain(
+        string memory assetSymbol,
+        uint256 amount,
+        uint256 chainId,
+        PaymentInfo.Payment memory payment
+    ) internal pure returns (uint256) {
+        // If action is paid for with tokens and the payment token is the transfer token, then add the payment max cost for the target chain to the amount needed
+        uint256 amountNeededOnChain = amount;
+        if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, assetSymbol)) {
+            amountNeededOnChain += PaymentInfo.findMaxCost(payment, chainId);
+        }
+
+        return amountNeededOnChain;
+    }
+
+    function needsBridgedFunds(
+        string memory assetSymbol,
+        uint256 amount,
+        uint256 chainId,
+        Accounts.ChainAccounts[] memory chainAccountsList,
+        PaymentInfo.Payment memory payment
+    ) internal pure returns (bool) {
+        uint256 balanceOnChain = getBalanceOnChain(assetSymbol, chainId, chainAccountsList, payment);
+        uint256 amountNeededOnChain = getAmountNeededOnChain(assetSymbol, amount, chainId, payment);
+
+        return balanceOnChain < amountNeededOnChain;
```

Comment:
These diffs are just pulling out the logic in `needsBridgedFunds` into two separate helper functions.